"""
AIè§’è‰²æ‰®æ¼”gRPCæœåŠ¡ç«¯
æä¾›é€šè¿‡gRPCè°ƒç”¨AIè§’è‰²æ‰®æ¼”åŠŸèƒ½çš„èƒ½åŠ›
"""

import grpc
from concurrent import futures
import time
import sys
import os

# æ·»åŠ å½“å‰ç›®å½•å’Œä¸Šçº§ç›®å½•åˆ°Pythonè·¯å¾„
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# å¯¼å…¥ç”Ÿæˆçš„gRPCä»£ç 
import role_play_pb2
import role_play_pb2_grpc

# å¯¼å…¥è§’è‰²æ‰®æ¼”åŠŸèƒ½
from role_play import PRESET_ROLES, set_role, chat_with_role, reset_conversation, llm

class RolePlayServiceServicer(role_play_pb2_grpc.RolePlayServiceServicer):
    """è§’è‰²æ‰®æ¼”æœåŠ¡å®ç°"""
    
    def SetRole(self, request, context):
        """è®¾ç½®è§’è‰²"""
        try:
            # è°ƒç”¨åŸæœ‰çš„è®¾ç½®è§’è‰²å‡½æ•°
            result, _ = set_role(
                request.role_name,
                request.custom_name,
                request.custom_prompt,
                request.custom_greeting
            )
            
            # è¿”å›å“åº”
            return role_play_pb2.SetRoleResponse(
                success=True,
                message=result,
                greeting=""  # greetingå·²ç»åœ¨resultä¸­åŒ…å«äº†
            )
        except Exception as e:
            return role_play_pb2.SetRoleResponse(
                success=False,
                message=f"è®¾ç½®è§’è‰²å¤±è´¥: {str(e)}",
                greeting=""
            )
    
    def ChatWithRole(self, request, context):
        """ä¸è§’è‰²å¯¹è¯"""
        try:
            # è°ƒç”¨åŸæœ‰çš„å¯¹è¯å‡½æ•°
            response = chat_with_role(request.message, [])
            
            # æ£€æŸ¥æ˜¯å¦æ˜¯é”™è¯¯ä¿¡æ¯
            if response.startswith("âŒ") or response.startswith("âš ï¸"):
                return role_play_pb2.ChatResponse(
                    success=False,
                    response="",
                    error=response
                )
            
            return role_play_pb2.ChatResponse(
                success=True,
                response=response,
                error=""
            )
        except Exception as e:
            return role_play_pb2.ChatResponse(
                success=False,
                response="",
                error=f"å¯¹è¯å¤±è´¥: {str(e)}"
            )
    
    def ResetConversation(self, request, context):
        """é‡ç½®å¯¹è¯"""
        try:
            # è°ƒç”¨åŸæœ‰çš„é‡ç½®å‡½æ•°
            reset_conversation()
            
            return role_play_pb2.ResetResponse(
                success=True,
                message="å¯¹è¯å·²é‡ç½®"
            )
        except Exception as e:
            return role_play_pb2.ResetResponse(
                success=False,
                message=f"é‡ç½®å¯¹è¯å¤±è´¥: {str(e)}"
            )
    
    def GetPresetRoles(self, request, context):
        """è·å–é¢„è®¾è§’è‰²åˆ—è¡¨"""
        try:
            roles = []
            for emoji_name, role_data in PRESET_ROLES.items():
                # ä»"ğŸ§™â€â™‚ï¸ æ™ºæ…§å¯¼å¸ˆ"è¿™æ ·çš„å­—ç¬¦ä¸²ä¸­æå–emojiå’Œåç§°
                parts = emoji_name.split(' ', 1)
                emoji = parts[0] if len(parts) > 1 else ""
                name = role_data.get("name", "")
                description = role_data.get("description", "")
                
                roles.append(role_play_pb2.PresetRole(
                    name=name,
                    description=description,
                    emoji=emoji
                ))
            
            return role_play_pb2.GetPresetRolesResponse(roles=roles)
        except Exception as e:
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"è·å–é¢„è®¾è§’è‰²å¤±è´¥: {str(e)}")
            return role_play_pb2.GetPresetRolesResponse()

def serve():
    """å¯åŠ¨gRPCæœåŠ¡å™¨"""
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    role_play_pb2_grpc.add_RolePlayServiceServicer_to_server(
        RolePlayServiceServicer(), server
    )
    server.add_insecure_port('[::]:50051')
    server.start()
    print("AIè§’è‰²æ‰®æ¼”gRPCæœåŠ¡å™¨å·²å¯åŠ¨ï¼Œç›‘å¬ç«¯å£50051...")
    print("æŒ‰Ctrl+Cåœæ­¢æœåŠ¡å™¨")
    
    try:
        while True:
            time.sleep(86400)  # ä¸€å¤©
    except KeyboardInterrupt:
        server.stop(0)
        print("\næœåŠ¡å™¨å·²åœæ­¢")

if __name__ == '__main__':
    serve()